chia-client/src/peer.rs

```diff
@@ -0,0 +1,207 @@
+                                let cursor = &mut Cursor::new(message.data.as_ref());
+                                if let Ok(body) = NewPeakWallet::parse(cursor) {
+                                    event_sender.send(PeerEvent::NewPeakWallet(body)).ok();
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+            }
+        });
+
+        Self {
+            sink: Mutex::new(sink),
+            inbound_task,
+            event_receiver,
+            requests,
+            nonce: Mutex::new(0),
+        }
+    }
+
+    pub async fn perform_handshake(&self, network_id: String, node_type: NodeType) -> Result<()> {
+        let handshake = Handshake {
+            network_id,
+            protocol_version: "0.0.34".to_string(),
+            software_version: "0.0.0".to_string(),
+            server_port: 0,
+            node_type,
+            capabilities: vec![
+                (1, "1".to_string()),
+                (2, "1".to_string()),
+                (3, "1".to_string()),
+            ],
+        };
+        self.send(handshake).await
+    }
+
+    pub async fn send<T>(&self, body: T) -> Result<()>
+    where
+        T: Streamable + ChiaProtocolMessage,
+    {
+        // Create the message.
+        let message = Message {
+            msg_type: T::msg_type(),
+            id: None,
+            data: {
+                let mut body_bytes = Vec::new();
+                body.stream(&mut body_bytes)?;
+                body_bytes.into()
+            },
+        };
+
+        // Convert message to bytes.
+        let mut bytes = Vec::new();
+        message.stream(&mut bytes)?;
+
+        // Send the message through the websocket.
+        let mut sink = self.sink.lock().await;
+        sink.send(bytes.into()).await?;
+
+        Ok(())
+    }
+
+    pub async fn request<T, R>(&self, body: T) -> Result<R>
+    where
+        T: Streamable + ChiaProtocolMessage,
+        R: Streamable + ChiaProtocolMessage,
+    {
+        // Get the current nonce.
+        let mut nonce = self.nonce.lock().await;
+        let message_id = *nonce;
+
+        // Create the message.
+        let message = Message {
+            msg_type: T::msg_type(),
+            id: Some(message_id),
+            data: {
+                let mut body_bytes = Vec::new();
+                body.stream(&mut body_bytes)?;
+                body_bytes.into()
+            },
+        };
+
+        // Convert message to bytes.
+        let mut bytes = Vec::new();
+        message.stream(&mut bytes)?;
+
+        // Create a saved oneshot channel to receive the response.
+        let (sender, receiver) = oneshot::channel::<Message>();
+        self.requests.lock().await.insert(message_id, sender);
+
+        // Send the message.
+        if let Err(error) = self.sink.lock().await.send(bytes.into()).await {
```

Comment:
I did this just in case the response is somehow processed before the next line of the function after the request has been sent. I don't want to somehow miss the response and have it hang indefinitely since the oneshot channel wasn't in the hashmap at the point it was received. If you think this will never happen, the code can be a lot simpler.

Reply:
I'm not entirely sure how this code is supposed to be put together and work with coroutines. the way we do it in python is that every request is a coroutine (or rather, if you want to pipeline requests, you have to use multiple coroutines). There we have a main coroutine per peer that triages incoming messages to the coroutine that's waiting for that message. this is where the transaction IDs come in, to know where to send it.

but in that model, there's a kind of driver that does the actual sending and receiving of messages.